nncf/common/quantization/quantizer_propagation/solver.py

```diff
@@ -0,0 +1,1211 @@
+from nncf.common.utils.backend import __nncf_backend__
+
+if __nncf_backend__ == 'Torch':
+    # pylint:disable=wildcard-import
+    # pylint:disable=unused-wildcard-import
+    from nncf.torch.graph.operator_metatypes import *
+    DEFAULT_QUANT_TRAIT_TO_OP_DICT = {
+        QuantizationTrait.INPUTS_QUANTIZABLE: [
+            Conv2dMetatype,
+            Conv3dMetatype,
+            ConvTranspose2dMetatype,
+            ConvTranspose3dMetatype,
+            DepthwiseConv2dSubtype,
+            DepthwiseConv3dSubtype,
+            LinearMetatype,
+            HardTanhMetatype,
+            TanhMetatype,
+            ELUMetatype,
+            PRELUMetatype,
+            LeakyRELUMetatype,
+            LayerNormMetatype,
+            GELUMetatype,
+            SigmoidMetatype,
+            AddMetatype,
+            MulMetatype,
+            DivMetatype,
+            ExpMetatype,
+            ErfMetatype,
+            MatMulMetatype,
+            MeanMetatype,
+            RoundMetatype,
+            PixelShuffleMetatype,
+            BatchNormMetatype,
+            AvgPool2dMetatype,
+            AvgPool3dMetatype
+        ],
+        QuantizationTrait.NON_QUANTIZABLE: [
+            SoftmaxMetatype
+        ],
+        QuantizationTrait.CONCAT: [
+            CatMetatype
+        ],
+        QuantizationTrait.OUTPUT_QUANTIZATION_AS_WEIGHTS: [
+            EmbeddingMetatype,
+            EmbeddingBagMetatype
+        ]
+    }  # type: Dict[QuantizationTrait, List[OperatorMetatype]]
+else:
+    DEFAULT_QUANT_TRAIT_TO_OP_DICT = {}  # TODO (vshampor): populate with TF metatypes or unify metatypes
+
+
+class TransitionStatus(Enum):
+    SHOULD_TRANSITION = 0
+    SHOULD_MERGE = 1
+    SHOULD_NOT_TRANSITION = 2
+    SHOULD_WAIT_FOR_MERGE = 3
+
+
+class PropagationStrategy(Enum):
+    # While propagating up through a downward-branching node:
+    # ... do not merge at all
+    DO_NOT_MERGE_BRANCH_FQS = 0
+
+    # ... only merge for exact configuration space matches
+    MERGE_IF_ALL_BRANCH_FQ_OPTIONS_SAME = 1
+
+    # ... merge common parts, and if a branch quantizer has options for
+    # narrowing (bitwidth/mode/per-channel/etc.) in addition to
+    # the common part, keep the quantizer on branch
+    MERGE_WITH_POTENTIAL_REQUANTIZATION = 2
+
+    # ... merge common config options into a single config space for the global FQ,
+    # do not merge if this is impossible for the current branching situation and given
+    # HW config file
+    MERGE_WITH_SINGLE_FQ_RESULT = 3
+
+
+class FinalizedQuantizationProposal:
+    def __init__(self, single_config_quantizer_setup: SingleConfigQuantizerSetup,
+                 quant_prop_graph: QuantizerPropagationStateGraph):
+        self.single_config_quantizer_setup = single_config_quantizer_setup
+        self._quant_prop_graph = quant_prop_graph
+
+    @property
+    def quant_prop_graph(self):
+        return self._quant_prop_graph
+
+
+class QuantizationProposal:
```

Comment:
Could you please add a class docstring and a docstring for all "public" methods?

Reply:
Done